pandas/core/generic.py

```diff
@@ -6001,19 +6001,25 @@ def fillna(
                 )
 
             elif isinstance(value, (dict, ABCSeries)):
+                result = self if inplace else self.copy()
+
                 if axis == 1:
-                    raise NotImplementedError(
-                        "Currently only can fill "
-                        "with dict/Series column "
-                        "by column"
-                    )
+                    # To access column base
+                    result = result.T
 
-                result = self if inplace else self.copy()
-                for k, v in value.items():
-                    if k not in result:
+                for i in range(result.columns.size):
+                    label = result.columns[i]
+
+                    if label not in value.keys():
                         continue
-                    obj = result[k]
-                    obj.fillna(v, limit=limit, inplace=True, downcast=downcast)
+
+                    result.iloc[:, i] = result.iloc[:, i].fillna(
+                        value[label], limit=limit, inplace=False, downcast=downcast
+                    )
+
+                if axis == 1:
+                    result = result.T
+
                 return result if not inplace else None
```

Comment:
this doesn't make any sense if we have transposed

rather do

self._update_inplace(result._data)

Reply:
@jreback 
Okay i change logic. but two ".T" operations need. because what i try to do is axis==1 and axis==0 both can handle. current ".fillna" can't handle duplicated columns.